Problem trying to download content

Experiencing problem downloading from storage node. We get around 4K of the file then connection is dropped:

simple test with wget:
```
wget https://storage-node-1.joystream.org/asset/v0/d084fcd9-c489-57ff-9816-9917aaec0e1e/ef9818a5b76c-4989-b347-31128ebba?download 
--2019-04-14 17:15:53--  https://storage-node-1.joystream.org/asset/v0/d084fcd9-c489-57ff-9816-9917aaec0e1e/ef9818a5b76c-4989-b347-31128ebba?download
Resolving storage-node-1.joystream.org (storage-node-1.joystream.org)... 80.85.86.213
Connecting to storage-node-1.joystream.org (storage-node-1.joystream.org)|80.85.86.213|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 74063411 (71M) [application/octet-stream]
Saving to: 'ef9818a5b76c-4989-b347-31128ebba?download.3'

ef9818a5b76c-4989-b347-31128ebba?download   0%[                                                                                    ]   4.00K  --.-KB/s    in 0.03s   

2019-04-14 17:15:54 (149 KB/s) - Connection closed at byte 4096. Retrying.
```


Debug logs on storage node:

```bash
  joystream:repository Opening /ef9818a5b76c-4989-b347-31128ebba with mode r +1ms
  joystream:repository Initialized hyperdrive storage at: /home/joystream/STORAGE/repos/d0/84/d084fcd9-c489-57ff-9816-9917aaec0e1e +6ms
  joystream:repository Reading 4100 Bytes from the stream. +15ms
  joystream:repository Detected file type is application/octet-stream +3ms
  joystream:util:ranges RangeSender: RangeSender {
  name: 'ef9818a5b76c-4989-b347-31128ebba.bin',
  type: 'application/octet-stream',
  size: 74063411,
  ranges: undefined,
  download: true,
  read_offset: 0,
  range_index: -1,
  range_boundary: undefined,
  handlers: {},
  opened: false } +1ms
  joystream:util:ranges No ranges, just send the whole file. +1ms
  joystream:util:ranges Open succeeded: ef9818a5b76c-4989-b347-31128ebba.bin application/octet-stream +1ms
  joystream:util:ranges End of stream. +6ms
  joystream:util:ranges Error: Error: stream.push() after EOF
    at readableAddChunk (/home/joystream/storage-node-joystream/node_modules/readable-stream/lib/_stream_readable.js:271:30)
    at Class.Readable.push (/home/joystream/storage-node-joystream/node_modules/readable-stream/lib/_stream_readable.js:245:10)
    at check (/home/joystream/storage-node-joystream/node_modules/from2/index.js:60:16)
    at Request._callback (/home/joystream/storage-node-joystream/node_modules/hyperdrive/index.js:471:7)
    at Request.callback (/home/joystream/storage-node-joystream/node_modules/random-access-storage/index.js:161:8)
    at onread (/home/joystream/storage-node-joystream/node_modules/random-access-file/index.js:81:31)
    at FSReqWrap.wrapper [as oncomplete] (fs.js:462:17) +4ms
_http_outgoing.js:470
    throw new ERR_HTTP_HEADERS_SENT('set');
    ^

Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:470:11)
    at ServerResponse.header (/home/joystream/storage-node-joystream/node_modules/express/lib/response.js:767:10)
    at ServerResponse.expressOpenAPISend [as send] (/home/joystream/storage-node-joystream/lib/joystream/middleware/validate_responses.js:34:13)
    at RangeSender.on_error (/home/joystream/storage-node-joystream/lib/joystream/util/ranges.js:171:43)
    at Class.emit (events.js:189:13)
    at readableAddChunk (/home/joystream/storage-node-joystream/node_modules/readable-stream/lib/_stream_readable.js:271:16)
    at Class.Readable.push (/home/joystream/storage-node-joystream/node_modules/readable-stream/lib/_stream_readable.js:245:10)
    at check (/home/joystream/storage-node-joystream/node_modules/from2/index.js:60:16)
    at Request._callback (/home/joystream/storage-node-joystream/node_modules/hyperdrive/index.js:471:7)
    at Request.callback (/home/joystream/storage-node-joystream/node_modules/random-access-storage/index.js:161:8)
```

Node:  v10.15.3
NPM: 6.4.1
yarn: 1.15.2

note: I'm running storage node  like so:

`node bin/cli`

because : `npm install -g` fails

```
joystream@localhost:~/storage-node-joystream$ npm install -g
npm WARN enoent ENOENT: no such file or directory, open '/home/joystream/storage-node-joystream/lib/joystream/package.json'

npm ERR! path /home/joystream/storage-node-joystream/lib/joystream/package.json
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open
npm ERR! enoent ENOENT: no such file or directory, open '/home/joystream/storage-node-joystream/lib/joystream/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 
```

--- a/lib/joystream/core/repository.js
+++ b/lib/joystream/core/repository.js
@@ -24,10 +24,14 @@ const file_type = require('file-type');
 const hyperdrive = require('hyperdrive');
 const events = require('events');
 
+const { ReadableStreamBuffer } = require('stream-buffers');
+
 const debug = require('debug')('joystream:repository');
 
 const fsresolve = require('joystream/util/fs/resolve');
 const fswalk = require('joystream/util/fs/walk');
+
+const CHUNK_SIZE = 64 * 1024;
 
 
 /*
@@ -179,29 +183,51 @@ class Repository extends events.EventEmitter
         const stream = this.archive.createReadStream(fname, { flags: `${mode}+`,
           encoding: 'binary' });
 
-        stream.once('readable', () => {
-          debug('Reading', file_type.minimumBytes, 'Bytes from the stream.');
-          stream.pause();
-          const chunk = stream.read(file_type.minimumBytes);
-          var data = chunk;
+        // Have a reader function that detects the file type.
+        var buf = Buffer.alloc(0);
+        const ft_detector = (chunk) => {
+          // Ensure chunk is a Buffer
           if (typeof chunk === 'string') {
-            data = Buffer.from(chunk);
+            chunk = Buffer.from(chunk);
           }
-          stream.unshift(chunk);
-          stream.resume();
+          debug('Got', chunk.length, chunk);
 
-          var ft = { mime: 'application/octet-stream' };
-          if (null != data) {
-            const res = file_type(data);
+          // Add chunk to our detection buffer.
+          buf = Buffer.concat([buf, chunk]);
+
+          // If we have sufficient data buffered, detect file type.
+          if (buf.length >= file_type.minimumBytes) {
+            stream.off('data', ft_detector);
+
+            debug('Read enough for file type detection.');
+            var ft = { mime: 'application/octet-stream' };
+            const res = file_type(buf);
             if (res != null) {
               ft = res;
             }
-          }
-          debug('Detected file type is', ft.mime);
+            debug('File type detected as', ft.mime);
 
-          cb(null, ft.mime, stream);
-          stream.emit('readable');
-        });
+            // Now put the detection buffer into a readable memory stream,
+            // pass that on, and 'pipe' to it.
+            console.log(stream.readableHighWaterMark);
+            const bufstream = new ReadableStreamBuffer({
+              chunkSize: CHUNK_SIZE,
+            });
+            bufstream.put(buf);
+
+            cb(null, ft.mime, bufstream);
+
+            stream.on('data', (data) => {
+              bufstream.put(data);
+            });
+            stream.on('end', (data) => {
+              bufstream.put(data);
+              bufstream.stop();
+            });
+          }
+        };
+
+        stream.on('data', ft_detector);
         return;
       }
       else {

--- a/lib/joystream/substrate/balances.js
+++ b/lib/joystream/substrate/balances.js
@@ -66,6 +66,14 @@ class BalancesApi extends IdentitiesApi
   }
 
   /*
+   * Return the base transaction fee.
+   */
+  async baseTransactionFee()
+  {
+    return await this.api.query.balances.transactionBaseFee();
+  }
+
+  /*
    * Transfer amount currency from one address to another. The sending
    * address must be an unlocked key pair!
    */

--- a/lib/joystream/substrate/roles.js
+++ b/lib/joystream/substrate/roles.js
@@ -82,7 +82,9 @@ class RolesApi extends BalancesApi
     if (_.isEqual(params.raw, new Null())) {
       throw new Error(`Role ${role} is not defined!`);
     }
-    const result = params.raw.min_stake.add(params.raw.entry_request_fee);
+    const result = params.raw.min_stake
+      .add(params.raw.entry_request_fee)
+      .add(await this.baseTransactionFee());
     return result;
   }
 

--- a/lib/joystream/util/ranges.js
+++ b/lib/joystream/util/ranges.js
@@ -186,9 +186,11 @@ class RangeSender
   {
     // Assume hiding the actual error is best, and default to 404.
     debug('Error:', err);
-    this.response.status(err.code || 404).send({
-      message: err.message || `File not found: ${this.name}`
-    });
+    if (!this.response.headersSent) {
+      this.response.status(err.code || 404).send({
+        message: err.message || `File not found: ${this.name}`
+      });
+    }
     if (this.end_callback) {
       this.end_callback(err);
     }

